src/software-device.cpp

```diff
@@ -254,12 +254,27 @@ namespace librealsense
     }
 
 
-    void software_sensor::set_metadata(rs2_frame_metadata_value key, rs2_metadata_type value)
+    void software_sensor::set_metadata( rs2_frame_metadata_value key, rs2_metadata_type value )
     {
+        auto range = _metadata_parsers->equal_range( key );
+        if( range.first == range.second )
+        {
+            if( int( key ) < 0 || int( key ) >= _metadata_map.size() )
+                throw invalid_value_exception( "invalid metadata key " + std::to_string( int( key ) ) );
+            // At this time (and therefore for backwards compatibility) no register_metadata is required for SW sensors,
+            // and metadata persists between frames (!!!!!!!) unless clear_metadata() is used...
+            _metadata_parsers->emplace_hint( range.second, key, std::make_shared< md_array_parser >( key ) );
```

Comment:
For DDS devices `make_shared` will be called repeatedly for every metadata field on every frame.
We should allocate the parsers only once at start up and reference them on each relevant frame.

Reply:
This is development.
For DDS I already put in a `erase_metadata` function that would make only unset if a value isn't present, instead of the current `clear_metadata`.
For this PR, it only inserts a "parse" if it wasn't there before, so only the first time this function is called per metadata.